### ISSUE TITLE
Compatibility with TF2

### ISSUE THREAD

**Issue by orsharir** [HUJI-Deep/FlowKet#6]
Now that TF2.0 has reached beta (and API freeze), we can start testing if our library is compatible with it. The idea isn't to leverage any of the new capabilities of TF2.0 (for instance, there's going to be a new and better support for multi-gpu upon TF2.0 release), only to check that our current code can be ran as is using TF2 backend.

This is a low priority task, that can be postponed after the other issues are resolved.

**Comment by noamwies** [comment]
@orsharir 
I finished adding support for tf2.3 , now all tests passed.
I will merge the tf2 branch and close this issue when the training of Heisenberg model will reach acceptable accuracy.

Note that currently ```tf.distribute.MirroredStrategy``` doesn't works with the new ```AccumulateGradientOptimizerWrapper``` but it has low priority since our largest model & batch size configuration easily fit into 4 GPU's memory.

Finally the sampling doesn't works with mixed precision policy, if you have any idea for workaround let me know.

**Comment by orsharir** [comment]
Great work!

Regard gradient accumulation, that's fine for now, though it should probably be easier to implement if you'll switch to using custom tf.function training.

Regarding mixed precision, do you mean that it doesn't work when you use Keras' mixed precision policy, or that the sampling doesn't work when using low-precision parameters?

Finally, I think that the real gain would be to have a single tf.function that does a full training step (sampling, energy computation, and parameters update step). If we'll have that, we could run it all on TPUs very efficiently, and possibly scale to very large TPU pods.

**Comment by noamwies** [comment]
@orsharir 
Regarding mixed precision, my meaning is the first one, if we can "disable" the policy just for the sampling it will be great since in any case the sampling is relativity fast (comparing to the local energy computations).

Regarding single ```tf.function``` it's interesting, the training loop itself is pretty straightforward, But it requires to implement the operators in tensorflow and that might be more complicated, i will give it a try

BTW the tensorflow 2 runtime has some overhead on the creation of the sampler.
After some improvement (uncommitted)  I manage to reduce it to 25-35 minutes comparing to 11 minutes in tensorflow 1.14 (on our standard architecture ) however the sampling itself is  112%-128% faster (it depends whatever I using ```@tensorflow.function(experimental_compile=True)```)

**Comment by orsharir** [comment]
I suggest we open a new issue for the slow sampler construction to keep track on it. I'm closing this main issue because we now support TF2.
